### PR TITLE
Bump mypy version to ~=1.13.0

### PR DESCRIPTION
**Pull request by unkcpz**
- Bump mypy version to `~=1.13.0`
- fixes all failed mypy checks.

### CODE REVIEW

**Comment by danielhollas** [comment]
I suspect that the sudden pre-commit failures are due to a new pydantic version 2.10 which was released yesterday.

https://docs.pydantic.dev/latest/changelog/

**Comment by unkcpz** [comment]
> I suspect that the sudden pre-commit failures are due to a new pydantic version 2.10 which was released yesterday.

Right, but the error didn't show up with the new mypy, so I think we just bump mypy and solve the new issues here.

**Comment by danielhollas** [comment]
It's not because of mypy, they actually released a patch version 2.10.1 that appear to fix some of the issues. But I agree that we might as well bump mypy.

**Comment by unkcpz** [comment]
Here are all errors fixed:

```
src/aiida/common/lang.py:99: error: Argument 1 has incompatible type "SelfType@__get__"; expected "SelfType@__init__"  [arg-type]
src/aiida/common/pydantic.py:45: error: Item "None" of "Any | None" has no attribute "metadata"  [union-attr]
src/aiida/manage/manager.py:433: error: Argument 1 to "Runner" has incompatible type "**dict[str, float | Any]"; expected "AbstractEventLoop | None"  [arg-type]
src/aiida/manage/manager.py:433: error: Argument 1 to "Runner" has incompatible type "**dict[str, float | Any]"; expected "bool"  [arg-type]
src/aiida/manage/manager.py:433: error: Argument 1 to "Runner" has incompatible type "**dict[str, float | Any]"; expected "Persister | None"  [arg-type]
src/aiida/plugins/utils.py:67: error: Subclass of "str" and "FunctionType" cannot exist: "FunctionType" is final  [unreachable]
src/aiida/plugins/utils.py:67: error: Subclass of "type" and "FunctionType" cannot exist: "FunctionType" is final  [unreachable]
src/aiida/orm/nodes/data/singlefile.py:74: error: Unused "type: ignore" comment  [unused-ignore]
src/aiida/orm/nodes/data/list.py:84: error: Signature of "pop" incompatible with supertype "MutableSequence"  [override]
src/aiida/orm/nodes/data/list.py:84: note:      Superclass:
src/aiida/orm/nodes/data/list.py:84: note:          def pop(self, index: int = ...) -> Any
src/aiida/orm/nodes/data/list.py:84: note:      Subclass:
src/aiida/orm/nodes/data/list.py:84: note:          def pop(self, **kwargs: Any) -> Any
src/aiida/orm/nodes/data/list.py:92: error: Signature of "index" incompatible with supertype "Sequence"  [override]
src/aiida/orm/nodes/data/list.py:92: note:      Superclass:
src/aiida/orm/nodes/data/list.py:92: note:          def index(self, value: Any, start: int = ..., stop: int = ...) -> int
src/aiida/orm/nodes/data/list.py:92: note:      Subclass:
src/aiida/orm/nodes/data/list.py:92: note:          def index(self, value: Any) -> Any
src/aiida/orm/nodes/data/array/projection.py:281: error: Signature of "set_orbitals" incompatible with supertype "OrbitalData"  [override]
src/aiida/orm/nodes/data/array/projection.py:281: note:      Superclass:
src/aiida/orm/nodes/data/array/projection.py:281: note:          def set_orbitals(self, orbitals: Any) -> Any
src/aiida/orm/nodes/data/array/projection.py:281: note:      Subclass:
src/aiida/orm/nodes/data/array/projection.py:281: note:          def set_orbitals(self, **kwargs: Any) -> Any
src/aiida/engine/processes/workchains/restart.py:160: error: Argument 1 to "validate_handler_overrides" has incompatible type "type[BaseRestartWorkChain]"; expected "BaseRestartWorkChain"  [arg-type]
src/aiida/engine/processes/calcjobs/calcjob.py:1065: error: Argument 1 to "asdict" has incompatible type "DataclassInstance | type[DataclassInstance]"; expected "DataclassInstance"  [arg-type]
tests/cmdline/groups/test_dynamic.py:20: error: Incompatible types in assignment (expression has type "bool", variable has type "str")  [assignment]
src/aiida/tools/pytest_fixtures/daemon.py:130: error: Right operand of "and" is never evaluated  [unreachable]
src/aiida/tools/pytest_fixtures/daemon.py:131: error: Statement is unreachable  [unreachable]
```

**Comment by unkcpz** [comment]
It seems last commit change fail the test.

File: .readthedocs.yml
Comment (unkcpz) [inline .readthedocs.yml:14]: Not sure, but it seems the `ModuleType` can not recognized when build from py3.11.

File: src/aiida/common/pydantic.py
Comment (unkcpz) [inline src/aiida/common/pydantic.py:44]: `Field` seems can return None, if so the `field_info` won't have `metadata` attribute.

File: src/aiida/engine/processes/calcjobs/calcjob.py
Comment (unkcpz) [inline src/aiida/engine/processes/calcjobs/calcjob.py:1065]: Not sure I should cast the type here, for the readability, I ignore the error.

File: tests/cmdline/groups/test_dynamic.py
Comment (unkcpz) [inline tests/cmdline/groups/test_dynamic.py:20]: It is test, so I just ignore it.

File: src/aiida/orm/nodes/data/array/projection.py
Comment (unkcpz) [inline src/aiida/orm/nodes/data/array/projection.py:281]: I'll just leave `kwargs` as it is and ignore type checking.

File: src/aiida/common/pydantic.py
Comment (danielhollas) [reply src/aiida/common/pydantic.py:44]: Nice fix. cc @edan-bainglass for visibility

File: src/aiida/engine/processes/workchains/restart.py
Comment (danielhollas) [inline src/aiida/engine/processes/workchains/restart.py:32]: Interesting. It might be worth adding a comment that explains why this is needed. What was the mypy error?

File: src/aiida/tools/pytest_fixtures/daemon.py
Comment (danielhollas) [inline src/aiida/tools/pytest_fixtures/daemon.py:119]: Why is `t.Any` needed here? This makes the type checking weaker, would be nice if we could avoid it.

File: src/aiida/orm/nodes/data/list.py
Comment (danielhollas) [inline src/aiida/orm/nodes/data/list.py:85]: This seems like potentially breaking change. Can you maybe add type ignore for now and then submit a separate PR where we can discuss more (it would be good to not block this PR to unblock pre-commit failures)

**Review by danielhollas — CHANGES_REQUESTED** [review]
Thanks @unkcpz! Can you revert the changes to `src/aiida/orm/nodes/data/list.py` and only add type: ignore, so we can merge this, and make a separate PR for those changes? Otherwise LGTM

File: src/aiida/orm/nodes/data/list.py
Comment (danielhollas) [reply src/aiida/orm/nodes/data/list.py:85]: For reference, I am copying the mypy error:

```python
src/aiida/orm/nodes/data/list.py:84: error: Signature of "pop" incompatible with supertype "MutableSequence"  [override]
src/aiida/orm/nodes/data/list.py:84: note:      Superclass:
src/aiida/orm/nodes/data/list.py:84: note:          def pop(self, index: int = ...) -> Any
src/aiida/orm/nodes/data/list.py:84: note:      Subclass:
src/aiida/orm/nodes/data/list.py:84: note:          def pop(self, **kwargs: Any) -> Any
src/aiida/orm/nodes/data/list.py:92: error: Signature of "index" incompatible with supertype "Sequence"  [override]
src/aiida/orm/nodes/data/list.py:92: note:      Superclass:
src/aiida/orm/nodes/data/list.py:92: note:          def index(self, value: Any, start: int = ..., stop: int = ...) -> int
src/aiida/orm/nodes/data/list.py:92: note:      Subclass:
src/aiida/orm/nodes/data/list.py:92: note:          def index(self, value: Any) -> Any
```

File: src/aiida/orm/nodes/data/array/projection.py
Comment (danielhollas) [reply src/aiida/orm/nodes/data/array/projection.py:281]: Can you add a TODO comment with the error? The mypy error suggests that this is an actual error? Or do you think it's a false positive?

```python
src/aiida/orm/nodes/data/array/projection.py:281: error: Signature of "set_orbitals" incompatible with supertype "OrbitalData"  [override]
src/aiida/orm/nodes/data/array/projection.py:281: note:      Superclass:
src/aiida/orm/nodes/data/array/projection.py:281: note:          def set_orbitals(self, orbitals: Any) -> Any
src/aiida/orm/nodes/data/array/projection.py:281: note:      Subclass:
src/aiida/orm/nodes/data/array/projection.py:281: note:          def set_orbitals(self, **kwargs: Any) -> Any
```

File: src/aiida/engine/processes/workchains/restart.py
Comment (unkcpz) [reply src/aiida/engine/processes/workchains/restart.py:32]: It is

```
src/aiida/engine/processes/workchains/restart.py:160: error: Argument 1 to "validate_handler_overrides" has incompatible type "type[BaseRestartWorkChain]"; expected "BaseRestartWorkChain"  [arg-type]
```

File: src/aiida/tools/pytest_fixtures/daemon.py
Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py:119]: Because there is an else branch otherwise the branch is unreachable.

File: src/aiida/engine/processes/workchains/restart.py
Comment (danielhollas) [reply src/aiida/engine/processes/workchains/restart.py:32]: Ah, I see, now that I think about it more it makes sense, it's a type of that class, not a class instance, that is being passed down.

File: src/aiida/tools/pytest_fixtures/daemon.py
Comment (danielhollas) [reply src/aiida/tools/pytest_fixtures/daemon.py:119]: Yes, makes sense, sorry I originally missed your comment with the errors.

I think the correct fix here is to add type ignores on those branches, since those essentially represent runtime type checking, which we need to enforce. At the same time we don't want to make static type checking weaker.

Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py:119]: I may not fully agree. I think types in the function signature plays two functionalities. On the one hand, it partially is the hint for the function behavior, like in this case in the function body it requires to explicitly deal with three types and raise if it doesn't know what to do with certain type. On the other hand, the type in the function signature is for other function to know what should be passed to the function when it is called. 
Ideally the `# type: ignore` should all be addressed. If we just rely on runtime type check, then the type should be `t.Any`.

File: tests/cmdline/groups/test_dynamic.py
Comment (GeigerJ2) [reply tests/cmdline/groups/test_dynamic.py:20]: Actually, when I run mypy 1.13 on this, the `# type: ignore[assignment]` is not necessary?

**Review by GeigerJ2 — COMMENTED** [review]
LGTM. Thanks for taking the time to look into this, @unkcpz!

**Review by GeigerJ2 — APPROVED** [review]
LGTM. Thanks for taking the time to look into this, @unkcpz!

**Comment by unkcpz** [comment]
> Actually, when I run mypy 1.13 on this, the # type: ignore[assignment] is not necessary?

This is the one required by the new pydantic. If you update the pydantic to 2.10.1 you can probably see this.

File: src/aiida/orm/nodes/data/list.py
Comment (unkcpz) [reply src/aiida/orm/nodes/data/list.py]: Moved to PR #6635

File: src/aiida/tools/pytest_fixtures/daemon.py
Comment (danielhollas) [reply src/aiida/tools/pytest_fixtures/daemon.py:119]: >  Ideally the # type: ignore should all be addressed.

I agree with this sentiment, but introducing `Any` is just another way to silence type checking. :-) I would argue that in this case `Any` is worse, since it "turns off" type checking of this parameter for the whole function, whereas the "type: ignore" would only target a specific part of the code. In this case, there is a real tension between static and runtime type checking: the else branches that are "unreachable" to the static type checker, but static types are not enforce at runtime in Python, and we can't enforce users to use mypy (e.g. in plugins that use this fixture), hence the additional runtime check.

> If we just rely on runtime type check, then the type should be t.Any

But we don't want to rely on runtime checks only. If we can catch it statically that's certainly better, no?

Sorry if I am being terse, I am teaching now, am happy to write more later today if needed.

Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py:119]: Right, I see your point now. I think you were suggest change to 

```
submittable: type[Process] | 'ProcessBuilder' | 'ProcessNode' # type: ignore
```

I thought you were suggest to not using type and just ignore it, sorry. I'll change this.

Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py]: Seems it is just works even without the `t.Any`. Please have a look.

Comment (danielhollas) [reply src/aiida/tools/pytest_fixtures/daemon.py]: Oh no, my suggestion was to use the type ignores where mypy needs them, which is on the else branches I believe.

Comment (danielhollas) [reply src/aiida/tools/pytest_fixtures/daemon.py]: Hmm, I don't think that's correct. :sweat_smile: 

I think the error goes away simply because you introduced a type that would fall through to the else branch. You can try that if you do this `Process | 'ProcessBuilder' | 'ProcessNode' | None` it would silence the error as well but would not be correct.

Comment (danielhollas) [reply src/aiida/tools/pytest_fixtures/daemon.py]: btw: Thanks for doing this, fixing type issues is a very annoying work (at least to me :-) )

Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py]: So you suggest to what? 

Should be
```
-> type[Process] | 'ProcessBuilder' | 'ProcessNode' # type: ignore
```
?

Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py]: > Thanks for doing this, fixing type issues is a very annoying work 

True, for this part, I'd much prefer the strict type language. Doing this in python actually much more tedious.

Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py]: But with change it to 
```
type[Process] | 'ProcessBuilder' | 'ProcessNode'
```

The else branch didn't ask for type ignore by mypy anymore. So I think with `t.Any` is correct and better.

Comment (danielhollas) [reply src/aiida/tools/pytest_fixtures/daemon.py:119]: I think the correct thing to do here is to NOT change the type annotation, since it was correct.

add type: ignore to the else branch, where the error originates.
The error is not really about type annotation, but about unreachable code. But we already know that branch is unreachable, unless the user screws up and passes a wrong type.

Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py:119]: `Process` -> `type[Process]` should be a correct change, in this function, it expect both workchain and calcjob which are subclass of `Process`. If it is just Process, then the `issubclass(submittable, Process)` part is useless because `and` shortcut the condition. 
We can have a zoom chat if you like, can be easier.

File: src/aiida/orm/nodes/data/array/projection.py
Comment (agoscinski) [reply src/aiida/orm/nodes/data/array/projection.py:281]: Can we just merge it and if this is a problem make an issue out of it?

File: src/aiida/engine/processes/calcjobs/calcjob.py
Comment (agoscinski) [reply src/aiida/engine/processes/calcjobs/calcjob.py:1065]: Would try to solve this in a subsequent PR

**Review by agoscinski — APPROVED** [review]
Please merge because it is blocking all PRs, we can put another PR to fix the remaining small issues.

File: src/aiida/tools/pytest_fixtures/daemon.py
Comment (unkcpz) [reply src/aiida/tools/pytest_fixtures/daemon.py]: In mega office, we decide to merge this one and continue the discussion here to not block other PRs. @danielhollas hope you don't mind.

Comment (danielhollas) [reply src/aiida/tools/pytest_fixtures/daemon.py]: Sure, go ahead

File: src/aiida/plugins/utils.py
Comment (danielhollas) [inline src/aiida/plugins/utils.py:41]: @unkcpz FYI I think this has a similar issue that we discussed, I don't think adding Any here is correct. Let's discuss on the other PR, but just flagging it here as well so we don't forget.

Comment (danielhollas) [reply src/aiida/plugins/utils.py:41]: The original error was:

> src/aiida/plugins/utils.py:67: error: Subclass of "str" and "FunctionType" cannot exist: "FunctionType" is final  [unreachable]

Still struggling to understand, will need to experiment a bit.

**Pull request merged**
